x/dex/cache/cache.go

```diff
@@ -49,103 +51,125 @@ type LiquidationRequests []LiquidationRequest
-	if _, ok := s.BlockCancels[contractAddr]; !ok {
-		s.BlockCancels[contractAddr] = map[typesutils.PairString]*BlockCancellations{}
-	}
-	if _, ok := s.BlockCancels[contractAddr][pair]; !ok {
-		emptyBlockCancels := BlockCancellations([]types.Cancellation{})
-		s.BlockCancels[contractAddr][pair] = &emptyBlockCancels
-	}
-	return s.BlockCancels[contractAddr][pair]
+	cancelsForContract, _ := s.BlockCancels.LoadOrStore(contractAddr, &sync.Map{})
+	emptyBlockCancels := BlockCancellations([]types.Cancellation{})
+	cancelsForPair, _ := cancelsForContract.(*sync.Map).LoadOrStore(pair, &emptyBlockCancels)
+	return cancelsForPair.(*BlockCancellations)
 }
 
 func (s *MemState) GetDepositInfo(contractAddr typesutils.ContractAddress) *DepositInfo {
-	if _, ok := s.DepositInfo[contractAddr]; !ok {
-		s.DepositInfo[contractAddr] = NewDepositInfo()
-	}
-	return s.DepositInfo[contractAddr]
+	depositsForContract, _ := s.DepositInfo.LoadOrStore(contractAddr, NewDepositInfo())
+	return depositsForContract.(*DepositInfo)
 }
 
 func (s *MemState) GetLiquidationRequests(contractAddr typesutils.ContractAddress) *LiquidationRequests {
-	if _, ok := s.LiquidationRequests[contractAddr]; !ok {
-		emptyRequests := LiquidationRequests([]LiquidationRequest{})
-		s.LiquidationRequests[contractAddr] = &emptyRequests
-	}
-	return s.LiquidationRequests[contractAddr]
+	emptyRequests := LiquidationRequests([]LiquidationRequest{})
+	liquidationsForContract, _ := s.LiquidationRequests.LoadOrStore(contractAddr, &emptyRequests)
+	return liquidationsForContract.(*LiquidationRequests)
 }
 
 func (s *MemState) Clear() {
-	s.BlockOrders = map[typesutils.ContractAddress]map[typesutils.PairString]*BlockOrders{}
-	s.BlockCancels = map[typesutils.ContractAddress]map[typesutils.PairString]*BlockCancellations{}
-	s.DepositInfo = map[typesutils.ContractAddress]*DepositInfo{}
-	s.LiquidationRequests = map[typesutils.ContractAddress]*LiquidationRequests{}
+	s.BlockOrders = &sync.Map{}
+	s.BlockCancels = &sync.Map{}
+	s.DepositInfo = &sync.Map{}
+	s.LiquidationRequests = &sync.Map{}
+}
+
+func (s *MemState) ClearCancellationForPair(contractAddr typesutils.ContractAddress, pair typesutils.PairString) {
+	cancelsForContract, _ := s.BlockCancels.LoadOrStore(contractAddr, &sync.Map{})
+	emptyBlockCancels := BlockCancellations([]types.Cancellation{})
+	cancelsForContract.(*sync.Map).Store(pair, &emptyBlockCancels)
 }
 
 func (s *MemState) DeepCopy() *MemState {
 	copy := NewMemState()
-	for contractAddr, _map := range s.BlockOrders {
-		for pair, blockOrders := range _map {
+	s.BlockOrders.Range(func(key, val any) bool {
+		contractAddr := key.(typesutils.ContractAddress)
+		_map := val.(*sync.Map)
+		_map.Range(func(pkey, pval any) bool {
+			pair := pkey.(typesutils.PairString)
+			blockOrders := pval.(*BlockOrders)
 			for _, blockOrder := range *blockOrders {
 				copy.GetBlockOrders(contractAddr, pair).AddOrder(blockOrder)
 			}
```

Comment:
yeah good call. I'll make an update

Reply:
well since golang doesn't support passing type as parameters or generics, this is probably as far as it goes: https://github.com/sei-protocol/sei-chain/pull/159/files#diff-f8d8121f7fdd1b62bd88a2ffb5afeeb0b601ed29a813686e9d327f538720a514R102